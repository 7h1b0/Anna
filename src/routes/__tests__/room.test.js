const request = require('supertest');
const knex = require('../../knexClient');
const Room = require('../../modules/models/room');
const User = require('../../modules/models/user');
const app = require('../../index.js');

const initRooms = [
  {
    room_id: 1,
    description: 'this is a test',
    name: 'room_1',
  },
  {
    room_id: 2,
    description: 'this is a second test',
    name: 'room_2',
  },
];

const user = {
  user_id: 1,
  username: 'test',
  password: '$2a$10$4ftuQxquI/5NR3POJy.2O.DmscxoSdCBzUvlnX2iXGMxtpqhd3w6O', // anna
  token: '8e6a76928f76d23665f78ff3688ca86422d5',
};

describe('Rooms API', () => {
  beforeAll(async () => {
    await knex(Room.TABLE).truncate();
    await knex(User.TABLE).insert(user);
  });

  beforeEach(async () => {
    await knex(Room.TABLE).insert(initRooms);
  });

  afterEach(async () => {
    await knex(Room.TABLE).truncate();
  });

  afterAll(async () => {
    await knex(User.TABLE).truncate();
    await knex.destroy();
  });

  describe('/api/rooms', () => {
    describe('GET', () => {
      it('should return all rooms', async () => {
        const response = await request(app)
          .get('/api/rooms')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.body).toMatchSnapshot();
      });

      it('should return 401 when user is not authenticated', async () => {
        const response = await request(app)
          .get('/api/rooms')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBe(401);
      });
    });
  });

  describe('/api/rooms/:id', () => {
    describe('GET', () => {
      it('should return an room', async () => {
        const response = await request(app)
          .get('/api/rooms/1')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.body).toEqual({
          roomId: 1,
          description: 'this is a test',
          name: 'room_1',
        });
      });

      it('should return 401 when user is not authenticated', async () => {
        const response = await request(app)
          .get('/api/rooms/2')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBe(401);
      });

      it("should retun 404 when a room don't exist", async () => {
        const response = await request(app)
          .get('/api/rooms/23')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(404);
      });
    });

    describe('PATCH', () => {
      it('should update an room', async () => {
        const response = await request(app)
          .patch('/api/rooms/1')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            description: 'this is a second test',
            name: 'room_updated',
          });

        expect(response.status).toBe(204);

        const room = await knex(Room.TABLE)
          .select('*')
          .where('room_id', 1);

        expect(room[0]).toEqual({
          description: 'this is a second test',
          name: 'room_updated',
          room_id: 1,
        });
      });

      it('should return 401 when user is not authenticated', async () => {
        const response = await request(app)
          .patch('/api/rooms/1')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBe(401);
      });

      it('should return 400 when request is invalid', async () => {
        const response = await request(app)
          .patch('/api/rooms/1')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            name: 'test_update',
            fake: true,
          });

        expect(response.status).toBe(400);
      });

      it("should retun 404 when a room don't exist", async () => {
        const response = await request(app)
          .patch('/api/rooms/23')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            description: 'this is a second test',
            name: 'room_updated',
          });

        expect(response.status).toBe(404);
      });
    });

    describe('DELETE', () => {
      it('should delete an room', async () => {
        const response = await request(app)
          .delete('/api/rooms/1')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(204);

        const room = await knex(Room.TABLE)
          .select('*')
          .where('room_id', 1);

        expect(room).toHaveLength(0);
      });

      it('should return 401 when user is not authenticated', async () => {
        const response = await request(app)
          .delete('/api/rooms/1')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBe(401);
      });

      it("should retun 404 when a room don't exist", async () => {
        const response = await request(app)
          .delete('/api/rooms/23')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(404);
      });
    });
  });
});
