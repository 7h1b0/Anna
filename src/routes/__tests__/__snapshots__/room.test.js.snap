// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rooms API /api/rooms GET should return all rooms 1`] = `
Array [
  Object {
    "createdAt": "2018-01-01T00:00:00.000Z",
    "createdBy": 1,
    "description": "this is a test",
    "name": "room_1",
    "roomId": 1,
    "updatedAt": "2018-01-02T00:00:00.000Z",
  },
  Object {
    "createdAt": "2018-01-01T00:00:00.000Z",
    "createdBy": 1,
    "description": "this is a second test",
    "name": "room_2",
    "roomId": 2,
    "updatedAt": "2018-01-02T00:00:00.000Z",
  },
]
`;

exports[`Rooms API /api/rooms POST should create a new room 1`] = `
Object {
  "createdAt": "2018-05-05T00:00:00.000Z",
  "createdBy": 1,
  "description": "a description",
  "name": "Living Room",
  "roomId": 3,
  "updatedAt": "2018-05-05T00:00:00.000Z",
}
`;

exports[`Rooms API /api/rooms POST should create a new room 2`] = `
Array [
  RowDataPacket {
    "created_at": 2018-05-05T00:00:00.000Z,
    "created_by": 1,
    "description": "a description",
    "name": "Living Room",
    "room_id": 3,
    "updated_at": 2018-05-05T00:00:00.000Z,
  },
]
`;

exports[`Rooms API /api/rooms/:id GET should return an room 1`] = `
Object {
  "createdAt": "2018-01-01T00:00:00.000Z",
  "createdBy": 1,
  "description": "this is a test",
  "name": "room_1",
  "roomId": 1,
  "updatedAt": "2018-01-02T00:00:00.000Z",
}
`;

exports[`Rooms API /api/rooms/:id PATCH should update an room 1`] = `
Array [
  RowDataPacket {
    "created_at": 2018-01-01T00:00:00.000Z,
    "created_by": 1,
    "description": "this is a second test",
    "name": "room_updated",
    "room_id": 1,
    "updated_at": 2018-05-05T00:00:00.000Z,
  },
]
`;
